### ISSUE TITLE
"dtype" system property is ignored when a new Nd4j context is created

### ISSUE THREAD

**Issue by mbabadi** [deeplearning4j/nd4j#1113]
... or perhaps overwritten by the backend resources (see Nd4j.java line 5053 in master). The issue is present in 0.4-rc3.10. The issue is not major (and maybe it is intended). In any case, it would be nice to be able to set the default data type via system properties.

As a side note, I think setting the default data type to single-precision float could be troublesome for some users: I got completely bogus results while learning the parameters of a factor analysis model using the EM algorithm on a large dataset (1000 samples in a 250'000 dimensional space). Roundoff errors become serious when reducing 10^9 single-precision floats to few (e.g. averaging). Perhaps float precision does not affect training convolutional neural nets as much and this justifies the default choice? (e.g. softmax functions can reduce the propagation of error; I'm not experienced in this area though)

**Comment by raver119** [comment]
Re: dtype. This behaviour was changed with 0.4.0, right now dtype property is persistent through contexts. 
So it would be nice if you check this issue with current stable release.

Re: precision Nothing can stop you to define DOUBLE precision at first line of your app. Or HALF, after next major release. That's whole point of switchable data types.

**Comment by raver119** [comment]
Any news for 0.4.0 there?  :)

**Comment by mbabadi** [comment]
Re: dtype -- Just tested 0.4.0 -- the issue is fixed. It also warns if a precision lowering type conversion takes place, which is a very good thing.

Re: precision -- oh I totally agree about the existence of dtype flexibility in nd4j, which is great! my concern is about the choice of HALF as the default dtype. In many everyday applications, single-digit floats lead to bogus results without the user realizing it, e.g. in QR/LU decompositions of large matrices, spectral analysis, solving large linear systems, calculating covariances, etc.

**Comment by raver119** [comment]
Dont choose it, if your task do not allow it. It's not like we're forcing you, or something like that.

Some tasks allow that, some - do not. Since nVidia CUDA is supporting it now - we're adding support for it too. But it's up to you and your task, will you use it or not :)

**Comment by raver119** [comment]
P.s. And no, HALF isn't going to be default dtype.  It'll be optional, like DOUBLE, and available on CUDA only for now. Maybe for Intel MKL, but we'll investigate that later.
